spec/compiler/semantic/module_spec.cr

```diff
@@ -226,15 +226,183 @@ describe "Semantic: module" do
         end
       end
 
-      class Baz(X)
+      class Bar(U)
+        include Foo(self)
+      end
+
+      Bar(Int32).new.foo
+      ") { generic_class("Bar", int32).metaclass }
+  end
+
+  it "includes generic module with self, and inherits it" do
+    assert_type("
+      module Foo(T)
+        def foo
+          T
+        end
+      end
+
+      class Bar(U)
+        include Foo(self)
+      end
+
+      class Baz < Bar(Int32)
+      end
+
+      Baz.new.foo
+      ") { types["Baz"].metaclass }
```

Comment:
From analyzing the reason of #6547 I bisect the commit to the merge of #6504 which leads to this PR.

I'm having second thoughts on the semantic here. I think references to `self` as generic arguments should be eager and the return value of `Baz.new.foo` should be `Bar(Int32).class`.

From the spec in this line, if the generic module to be included `Foo(T)` defines an operation `do(t : T)`, I would expect that `Baz#do` would be callable with any `Bar(Int32)`. Otherwise an instance of `Bar(Int32)` is not substitutable with an instance of `Baz`, breaking the inheritance relationship between `Baz` and `Bar(Int32)`.

Maybe the way to achieve the effect that enum types are comparable with only them selfs and generating compile time errors is by using macro inherited hook. I think that with macro inherited hook is clear that the base type is not including the module.

Is there any other scenario where we would want the `self` in `Comparable(self)` to be lazy evaluated other than Enum?

Reply:
I think `self` as a type reference should be removed from the language. It's confusing because of that same reason: does `self` means the type where it's used, or as a substitute for subclasses? We can choose one or the other, but it'll be still be confusing.

For example, if we have:

```crystal
class Foo
  def method(other : self)
  end
end

class Bar < Foo
end
```

Which of the following should work:

```crystal
foo = Foo.new
bar = Bar.new

foo.method(foo)
foo.method(bar)
bar.method(foo)
bar.method(bar)
```

Does `self` means "Foo or any of its subclasses" or "the **actual** type that's being called"? If it's the later case, it's also complex to implement if we have a `Foo+` type.

If we disallow `self`, all of these ambiguities disappear, and we are forced to use `Foo`, which is very clear.

For `Enum`, maybe an `inherited` macro could be defined to `include Comparable({{@type}})`. Or this: https://play.crystal-lang.org/#/r/4ree

Reply:
I mean, this: https://play.crystal-lang.org/#/r/4ref

Reply:
To late to ask silly question. 

>If self would be gone as a type restriction, how would you refer to the type were the module is included and use it as a type restriction? 

you either write the module name or there is a generic argument to use.

So, yes, it seems we can remove `self` as type restriction.

Reply:
I'd like to use lazy `self` type for associated type like Rust's `Self` type. But it is hard for now.

However, it breaks OOP principles surely. This PR is my failure. Sorry 🙇 

`self` type should become simple alias for current type. But it is a bit confusing due to difference between normal `self` usage. So removing `self` type is good idea to me.

IMHO the right way to keep `Enum` type safety is using `inherted` as @asterite  said. `Enum` type itself does not have the ability of `Comparable`, but a type inhering `Enum` is just comparable.

Reply:
Actually, I'm thinking that it might be good to keep `self`, but it should always be an alias of the type where it is used, and not lazily solved, even in method restrictions. For example we have `def_equals` which generates a method with a `self` restriction. Otherwise I'm not sure how it can be implemented (I'm seeing in Struct that we use `is_a?(self)` so it's basically the same). And, well, self is DRY.

Reply:
Let's leave `self` as type restriction for DRY and in macros expanded up the hierarchy then.

I found a bit weird that the `is_a?(self)` expression. `self.is_a?(self)` is always true but those two `self` are evaluated to different things. I would expect either `self.is_a?(typeof(self))` or `self.is_a?(TheType)` depending on what is wanted. But we can review that later.

Reply:
I use `self` a lot to type methods in modules to be included, and want it to refer to the final type. Without lazy `self` type, we can't type these methods or restrict args, right?

Reply:
E.g. https://github.com/ysbaddaden/rome/blob/master/src/query.cr

Reply:
@ysbaddaden You can use generic type parameter and macro:

```crystal
module Foo(X)
  def foo: Array(X)
    [] of X
  end
end

macro include_foo
  include Foo({{@type}})
end

class Bar
  include_foo
end
```

@bcardiff Is this what you said above https://github.com/crystal-lang/crystal/pull/3972#discussion_r210797252 ?

Reply:
Oh, right, then I'd have `class Model; include Query({{@type}}); end`. Not nearly as pretty but that could do the trick, I guess.